test/upgrade-e2e/post_upgrade_test.go

```diff
@@ -217,13 +236,84 @@ func waitForDeployment(t *testing.T, ctx context.Context, controlPlaneLabel stri
 		desiredNumReplicas = *managerDeployment.Spec.Replicas
 	}, time.Minute, time.Second)
 
-	var managerPods corev1.PodList
 	t.Logf("Ensure the number of remaining pods equal the desired number of replicas (%d)", desiredNumReplicas)
 	require.EventuallyWithT(t, func(ct *assert.CollectT) {
+		var managerPods corev1.PodList
 		require.NoError(ct, c.List(ctx, &managerPods, client.MatchingLabelsSelector{Selector: deploymentLabelSelector}))
-		require.Len(ct, managerPods.Items, 1)
+		require.Len(ct, managerPods.Items, int(desiredNumReplicas))
 	}, time.Minute, time.Second)
-	return &managerPods.Items[0]
+}
+
+// findLeaderPod finds the pod that has acquired the leader lease by inspecting the Lease resource.
+// This is more reliable than checking logs as it directly queries the Kubernetes API for the lease holder.
+func findLeaderPod(ctx context.Context, controlPlaneLabel string) (*corev1.Pod, error) {
+	deploymentLabelSelector := labels.Set{"app.kubernetes.io/name": controlPlaneLabel}.AsSelector()
+
+	// Map component name to its LeaderElectionID
+	leaseNameMap := map[string]string{
+		"operator-controller": "9c4404e7.operatorframework.io",
+		"catalogd":            "catalogd-operator-lock",
+	}
+
+	leaseName, ok := leaseNameMap[controlPlaneLabel]
+	if !ok {
+		return nil, fmt.Errorf("unknown control plane label: %s", controlPlaneLabel)
+	}
+
+	// Retry logic: keep checking until we find a leader or context times out
+	ticker := time.NewTicker(2 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, fmt.Errorf("timeout waiting for leader election: %w", ctx.Err())
+		case <-ticker.C:
+			var managerPods corev1.PodList
+			if err := c.List(ctx, &managerPods, client.MatchingLabelsSelector{Selector: deploymentLabelSelector}); err != nil {
+				continue
+			}
+
+			if len(managerPods.Items) == 0 {
+				continue
+			}
+
+			// If there's only one pod, it must be the leader
+			if len(managerPods.Items) == 1 {
+				return &managerPods.Items[0], nil
+			}
+
+			// For multi-replica deployments, check the Lease resource to find the leader
+			var lease coordinationv1.Lease
```

Comment:
Given that we use `testify` library in this project, we could use `require.Eventually()` for polling instead of implementing our custom solution.